src/core/server/server.cc

```diff
@@ -84,6 +86,227 @@
+            });
+      },
+      DEBUG_LOCATION);
+}
+
+void Server::ListenerInterface::LogicalConnection::CancelDrainGraceTimer() {
+  work_serializer_.Run(
+      [self = Ref(DEBUG_LOCATION, "CancelDrainGraceTimer")]() {
+        self->drain_grace_timer_handle_cancelled_ = true;
+        if (self->drain_grace_timer_handle_ !=
+            EventEngine::TaskHandle::kInvalid) {
+          self->event_engine()->Cancel(self->drain_grace_timer_handle_);
+          self->drain_grace_timer_handle_ = EventEngine::TaskHandle::kInvalid;
+        }
+      },
+      DEBUG_LOCATION);
+}
+
+void Server::ListenerInterface::LogicalConnection::OnDrainGraceTimer() {
+  work_serializer_.Run(
+      [self = Ref(DEBUG_LOCATION, "OnDrainGraceTimer")]() mutable {
+        // If the drain_grace_timer_ was not cancelled, disconnect
+        // immediately.
+        bool disconnect_immediately = false;
+        {
+          self->drain_grace_timer_handle_ = EventEngine::TaskHandle::kInvalid;
+          if (!self->drain_grace_timer_handle_cancelled_) {
+            disconnect_immediately = true;
+          }
+        }
+        if (disconnect_immediately) {
+          self->DisconnectImmediatelyImplLocked();
+        }
+      },
+      DEBUG_LOCATION);
+}
+
+//
+// Server::ListenerInterface
+//
+
+void Server::ListenerWrapper::Start() {
+  if (IsServerListenerEnabled()) {
+    if (server_->config_fetcher() != nullptr) {
+      auto watcher =
+          std::make_unique<ListenerWrapper::ConfigFetcherWatcher>(this);
+      config_fetcher_watcher_ = watcher.get();
+      server_->config_fetcher()->StartWatch(
+          grpc_sockaddr_to_string(listener_->resolved_address(), false).value(),
+          std::move(watcher));
+    } else {
+      {
+        MutexLock lock(&mu_);
+        started_ = true;
+        is_serving_ = true;
+      }
+      listener_->Start();
+    }
+  } else {
+    listener_->Start();
+  }
+}
+
+void Server::ListenerWrapper::Stop() {
+  if (IsServerListenerEnabled()) {
+    absl::flat_hash_set<OrphanablePtr<ListenerInterface::LogicalConnection>>
+        connections;
+    {
+      MutexLock lock(&mu_);
+      // Orphan the connections so that they can start cleaning up.
+      connections = std::move(connections_);
+      is_serving_ = false;
+    }
+    if (config_fetcher_watcher_ != nullptr) {
+      CHECK_NE(server_->config_fetcher(), nullptr);
+      server_->config_fetcher()->CancelWatch(config_fetcher_watcher_);
+    }
+  }
+  GRPC_CLOSURE_INIT(&destroy_done_, ListenerDestroyDone, server_,
+                    grpc_schedule_on_exec_ctx);
+  listener_->SetOnDestroyDone(&destroy_done_);
+  listener_.reset();
+}
+
+absl::optional<ChannelArgs> Server::ListenerWrapper::AddLogicalConnection(
+    OrphanablePtr<ListenerInterface::LogicalConnection> connection,
+    const ChannelArgs& args, grpc_endpoint* endpoint) {
+  RefCountedPtr<grpc_server_config_fetcher::ConnectionManager>
+      connection_manager;
+  {
+    MutexLock lock(&mu_);
```

Comment:
Yeah, that's a good point.  I'm just worried about the possibility of bugs here, since this is more complicated than just holding the lock the whole time.  I think we're currently handling it right by checking whether the connection manager has changed when we re-acquire the lock, but I'm concerned about other future bugs that could pop up here.

I guess for now, let's just add a comment explaining why we're doing this and what we need to be careful of whenever we change this code.  In particular, please call out the race that we're checking for between this and `ConfigFetcherWatcher::UpdateConnectionManager()`.